### PR TITLE
Add barabasi_albert_graph random graph functions

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds new random graph functions to rustworkx and rustworkx-core to implement a random graph generator using the Barabási–Albert preferential attachment method. It takes an input graph (defaulting to a star graph) and then extends it to a given size.

### CODE REVIEW

**Comment by enavarro51** [comment]
I'll take a look at this.

File: rustworkx-core/src/generators/random_graph.rs [hunk 1]
```diff
@@ -395,10 +401,129 @@ where
...
+    G: Data<NodeWeight = T, EdgeWeight = M>
+        + NodeIndexable
+        + GraphProp
+        + NodeCount
+        + Build
+        + Create,
+    for<'b> &'b G: GraphBase<NodeId = G::NodeId> + IntoEdgesDirected + IntoNodeIdentifiers,
+    F: FnMut() -> T,
+    H: FnMut() -> M,
+    G::NodeId: Eq + Hash + std::fmt::Debug,
```
Comment (enavarro51) [inline rustworkx-core/src/generators/random_graph.rs:469]: Is there a reason for the Debug here?

File: src/random_graph.rs [hunk 2]
```diff
@@ -379,3 +379,125 @@ pub fn random_geometric_graph(
...
     Ok(graph)
 }
+
+/// Generate a random graph using Barabási–Albert preferential attachment
+///
+/// A graph is grown to $n$ nodes by adding new nodes each with $m$ edges that
+/// are preferentially attached to existing nodes with high degree. All the edges
+/// and nodes added to this graph will have weights of ``None``.
+///
+/// The algorithm performed by this function are described in:
```
Comment (enavarro51) [inline src/random_graph.rs:389]:
```suggestion
/// The algorithm performed by this function is described in:
```

File: src/random_graph.rs [hunk 3]
```diff
@@ -379,3 +379,125 @@ pub fn random_geometric_graph(
...
+///
+/// A. L. Barabási and R. Albert "Emergence of scaling in random networks",
+/// Science 286, pp 509-512, 1999.
+///
+/// :param int n: The number of nodes to extend the graph to.
+/// :param int m: The number of edges to attach from a new node to existing nodes.
+/// :param int seed: An optional seed to use for the random number generator
```
Comment (enavarro51) [inline src/random_graph.rs:396]:
```suggestion
/// :param int seed: An optional seed to use for the random number generator.
```

File: src/random_graph.rs [hunk 4]
```diff
@@ -379,3 +379,125 @@ pub fn random_geometric_graph(
...
+        return Err(PyValueError::new_err("m must be > 0"));
+    }
+    if m >= n {
+        return Err(PyValueError::new_err("m must be < n"));
+    }
+    let graph = match core_generators::barabasi_albert_graph(
+        n,
+        m,
+        seed,
+        initial_graph.map(|x| x.graph),
```
Comment (enavarro51) [inline src/random_graph.rs:423]: Not exactly sure what this line is doing.

File: src/random_graph.rs [hunk 5]
```diff
@@ -379,3 +379,125 @@ pub fn random_geometric_graph(
...
+
+/// Generate a random graph using Barabási–Albert preferential attachment
+///
+/// A graph is grown to $n$ nodes by adding new nodes each with $m$ edges that
+/// are preferentially attached to existing nodes with high degree. All the edges
+/// and nodes added to this graph will have weights of ``None``. For the purposes
+/// of the extension algorithm all edges are treated as weak (meaning directionality
+/// isn't considered).
+///
+/// The algorithm performed by this function are described in:
```
Comment (enavarro51) [inline src/random_graph.rs:450]:
```suggestion
/// The algorithm performed by this function is described in:
```

**Review by enavarro51 — CHANGES_REQUESTED** [review]
A few minor changes and a couple of questions.

File: rustworkx-core/src/generators/random_graph.rs (hunk 1 above)
Comment (mtreinish) [reply rustworkx-core/src/generators/random_graph.rs:469]: There isn't this is a leftover from when I was debugging the code. I'll remove it

File: src/random_graph.rs (hunk 4 above)
Comment (mtreinish) [reply src/random_graph.rs:423]: `Option<T>.map()` is just syntactical sugar and this is the equivalent of

```rust
match initial_graph {
    Some(graph) => Some(graph.graph),
    None => None,
}
```

Basically the type for `initial_graph` is `Option<PyGraph>` but the rustworkx-core function needs `Option<StablePyGraph>` so I'm using a map here to pass the inner `.graph` attribute if initial graph is set.

File: rustworkx-core/src/generators/random_graph.rs (hunk 1 above)
Comment (mtreinish) [reply rustworkx-core/src/generators/random_graph.rs:469]:
```suggestion
    G::NodeId: Eq + Hash,
```

File: src/random_graph.rs (hunk 4 above)
Comment (enavarro51) [reply src/random_graph.rs:423]: Ah, got it. Thanks.

**Review by enavarro51 — APPROVED** [review]
LGTM.

**Pull request merged**